File: subject-set.py
import csv
from panoptes_client import Project, SubjectSet

# update this list of project ids
# to build linked subject sets data files
# that will be included in the API
project_ids = [ 12268 ]

for project in project_ids:
    for subject_set in project.links.subject_sets:
        print(subject_set.id)
        if 'indexFields' in subject_set.metadata:
            print(subject_set.metadata['indexFields'])
            index_fields = (subject_set.metadata['indexFields']).split(',')

            headers = ['priority', 'subject_id']
            for field in index_fields:
                headers.append(field)

            rows = []
            for subject in subject_set.subjects:
                row = {}
                row['subject_id'] = subject.id
                row['priority'] = -1
                if 'priority' in subject.metadata:
                    row['priority'] = subject.metadata['priority']
                if '#priority' in subject.metadata:
                    row['priority'] = subject.metadata['#priority']
                for field in index_fields:
                    row[field] = subject.metadata[field]
                rows.append(row)

            filename = 'data/{id}.csv'.format(id=subject_set.id)
            with open(filename, mode='w') as csv_file:
                writer = csv.DictWriter(csv_file, fieldnames=headers)
                writer.writeheader()

                for row in rows:
                    writer.writerow(row)

File: src/subject-set.js
const fs = require('fs')
const { unparse } = require('papaparse')

const fetchWithRetry = require('./fetchWithRetry')

const PROJECT_IDS = [ 24686 ]
const PAGE_SIZE = 100

function subjectMetadataRow(subject, indexFields = []) {
  const row = {
    subject_id: subject.id,
    priority: -1
  }
  row.priority = subject.metadata.priority || subject.metadata['#priority'] || -1
  indexFields.forEach(field => {
    row[field] = subject.metadata[field]
  })
  return row
}

async function getPagedSubjects(subjectSet, page = 1) {
  const { id } = subjectSet
  const indexFields = subjectSet.metadata.indexFields.split(',')
  const { subjects, meta } = await fetchWithRetry('/subjects', {
    subject_set_id: id,
    page_size: PAGE_SIZE,
    page
  })
  const rows = subjects.map(subject => subjectMetadataRow(subject, indexFields))
  if (meta.subjects.page_count > page) {
    const nextPage = await getPagedSubjects(subjectSet, page + 1)
    return rows.concat(nextPage)
  } else {
    console.log({ id: subjectSet.id, subjects: rows.length })
    return rows
  }
}

async function getPagedSubjectSets(project, page = 1) {
  const ids = project.links.subject_sets
  const response = await fetchWithRetry('/subject_sets', {
    id: ids.join(','),
    page_size: PAGE_SIZE,
    page
  })
  const { subject_sets, meta } = response
  if (meta.subject_sets.page_count > page) {
    const nextPage = await getPagedSubjectSets(project, page + 1)
    return subject_sets.concat(nextPage)
  } else {
    return subject_sets
  }
}

async function getSubjectSets(project) {
  const subjectSets = await getPagedSubjectSets(project)
  const indexedSets = subjectSets.filter(s => !!s.metadata.indexFields)
  console.log({ id: project.id, sets: indexedSets.length })
  return indexedSets
}

async function getProjects() {
  const { projects } = await fetchWithRetry('/projects', {
    id: PROJECT_IDS.join(',')
  })
  return projects
}

function onFileWrite(error) {
  if (error) {
    console.error(error)
  }
}

async function writeCSVFile(subjectSet) {
  const subjects = await getPagedSubjects(subjectSet)
  const cleanSubjects = [...new Set(subjects)]
  const csv = unparse(cleanSubjects)
  if (cleanSubjects?.length > 0) {
    fs.writeFile(`./data/subjects/${subjectSet.id}.csv`, csv, onFileWrite)
  }
  return cleanSubjects
}

async function buildDataFiles() {
  let subjectSets = []
  let count = 0
  const projects = await getProjects()
  const projectSubjectSets = await Promise.all(projects.map(getSubjectSets))
  projectSubjectSets.forEach(projectSets => {
    subjectSets = subjectSets.concat(projectSets)
  })
  const subjectSetSubjects = await Promise.all(subjectSets.map(writeCSVFile))
  subjectSetSubjects.forEach(subjects => {
    count = count + subjects.length
  })
  console.log('subject sets:', subjectSets.length)
  console.log('subjects:', count)
}

buildDataFiles()
